Add PulumiTV content and transcripts (#864)
pulumi/pulumi-hugo: themes/default/content/resources/building-container-based-aws-lambda-functions-in-python/index.md

    So, yeah, we create this bucket that's going to host our videos when we upload them, and then we're going to create a repository for hosting our container images. So that's all these first two lines are doing. And then we grab an authorization token from ECR and we use that to feed into this image resource. And what this image resource is doing is you can see it has a build parameter, and this is taking... This is a context we're going to build. So later on I'll show you this Docker context, this exact same one I showed last week, but I'll show briefly here as well. And then we give it the image name, which is just going to be the URL of the repo, and then the registry information. So here, this is where I need the credentials, where I need the username and the passwords that we can actually Docker log in to push our image.

>     After that, you kind of... At this point of the program, what will have happened after we run Pulumi app is we'll have created the bucket, we'll have created the repo, and then we will have pushed this image into that repo. Next, we want to create the function that's based on this image. So we create a role and a policy attachment for that role. So you can see we're using very simple AssumeRole policy here, and then actually here we're kind of showing off some of the new enum support we have across all our languages. So this enum is projected into all of the languages and it's available in C# and Go and Python here and obviously Node as well.

Reviewer: ```suggestion
    After that, you kind of... At this point of the program, what will have happened after we run 'pulumi up' is we'll have created the bucket, we'll have created the repo, and then we will have pushed this image into that repo. Next, we want to create the function that's based on this image. So we create a role and a policy attachment for that role. So you can see we're using very simple AssumeRole policy here, and then actually here we're kind of showing off some of the new enum support we have across all our languages. So this enum is projected into all of the languages and it's available in C# and Go and Python here and obviously Node as well.
```

    And so here you can see we give it this function ARN and then we give it this event. This is for object creation because obviously we care about any videos that get created, and then we filter on this MP4 suffix since we're looking for specifically video files. Just because it's easier, we actually write JPEGs into the same bucket. So we obviously want to make sure we're not infinitely recursing on ourselves. And so we're just looking at the video files. And this depends on having this permission obviously. So we have this explicit dependency we define here.

>     So that's how this all gets this wired up. And then we finally export the bucket name for convenience. And so I actually ran this earlier. So I ran 'pulumi up'. It provided all the resources, shoved them all into AWS. And when I run this copy of the sample file into the bucket, if I show the output, this is running Pulumi logs, I'm just tailing the logs from CloudWatch. And you can see here, this is the thumbnail or task or function that's getting invoked. And here it's copying the file from S3 into the temporary directory on the container, and then it's running FFmpeg to extract that image. So you can actually see this thumbnails is getting saved here in the sample that JPEG. So you can see here it's copying as well. So let's actually copy this locally and see that it actually worked. And if we open up that sample, there's the thumbnail. And this is a video of a waterfall. So great success. It all worked.

Reviewer: ```suggestion
    So that's how this all gets this wired up. And then we finally export the bucket name for convenience. And so I actually ran this earlier. So I ran 'pulumi up'. It provided all the resources, shoved them all into AWS. And when I run this copy of the sample file into the bucket, if I show the output, this is running Pulumi logs, I'm just tailing the logs from CloudWatch. And you can see here, this is the thumbnail or task or function that's getting invoked. And here it's copying the file from S3 into the temporary directory on the container, and then it's running FFmpeg to extract that image. So you can actually see this thumbnail is getting saved here in the sample that JPEG. So you can see here it's copying as well. So let's actually copy this locally and see that it actually worked. And if we open up that sample, there's the thumbnail. And this is a video of a waterfall. So great success. It all worked.
```

>     I mentioned I would take a quick look at the Docker built context here. And this is just a Docker file. Exact same one as last time. Didn't touch anything here. So actually one of the cool things I'm showing here is that we authored our Pulumi program in Python, but our Lambda function and the Docker file that you can see is building off this Node.js base Lambda image. So actually the function is written... The Lambda function itself is still written in Node. And so here, the critical line here, we're installing FFmpeg and then copying it, copying this actual function handler. And the function handler is very simple. It's really just shelling out to copy to S3, running FFmpeg, and then copying the result of the thumbnail file back into S3.

Reviewer: ```suggestion
    I mentioned I would take a quick look at the Docker build context here. And this is just a Docker file. Exact same one as last time. Didn't touch anything here. So actually one of the cool things I'm showing here is that we authored our Pulumi program in Python, but our Lambda function and the Docker file that you can see is building off this Node.js base Lambda image. So actually the function is written... The Lambda function itself is still written in Node. And so here, the critical line here, we're installing FFmpeg and then copying it, copying this actual function handler. And the function handler is very simple. It's really just shelling out to copy to S3, running FFmpeg, and then copying the result of the thumbnail file back into S3.
```
